cli/run_util.go

```diff
@@ -42,440 +42,460 @@ import (
 	}
 
-	return false, nil
+	return buildRunResults
 }
 
-func registerPrMode(isPRMode bool) error {
-	configs.IsPullRequestMode = isPRMode
-	return os.Setenv(configs.PRModeEnvKey, strconv.FormatBool(isPRMode))
+type activateAndRunStepResult struct {
+	Step               stepmanModels.StepModel
+	StepInfoPtr        stepmanModels.StepInfoModel
+	StepRunStatus      models.StepRunStatus
+	StepRunExitCode    int
+	StepRunErr         error
+	PrintStepHeader    bool
+	RedactedStepInputs map[string]string
+	OutputEnvironments []envmanModels.EnvironmentItemModel
 }
 
-func isCIMode(ciGlobalFlagPtr *bool, inventoryEnvironments []envmanModels.EnvironmentItemModel) (bool, error) {
-	if ciGlobalFlagPtr != nil {
-		return *ciGlobalFlagPtr, nil
+func newActivateAndRunStepResult(step stepmanModels.StepModel, stepInfoPtr stepmanModels.StepInfoModel, stepRunStatus models.StepRunStatus, stepRunExitCode int, stepRunErr error, printStepHeader bool, redactedStepInputs map[string]string, outputEnvironments []envmanModels.EnvironmentItemModel) activateAndRunStepResult {
+	return activateAndRunStepResult{Step: step, StepInfoPtr: stepInfoPtr, StepRunStatus: stepRunStatus, StepRunExitCode: stepRunExitCode, StepRunErr: stepRunErr, PrintStepHeader: printStepHeader, RedactedStepInputs: redactedStepInputs, OutputEnvironments: outputEnvironments}
+}
+
+func (r WorkflowRunner) activateAndRunStep(
+	step stepmanModels.StepModel,
+	stepID string,
+	stepIDx int,
+	defaultStepLibSource string,
+	stepExecutionID string,
+	tracker analytics.Tracker,
+	environments []envmanModels.EnvironmentItemModel,
+	secrets []envmanModels.EnvironmentItemModel,
+	buildRunResults models.BuildRunResultsModel,
+	isStepLibOfflineMode bool,
+	containerID, groupID string,
+	stepStartTime time.Time,
+	stepStartedProperties coreanalytics.Properties,
+) activateAndRunStepResult {
+	//
+	// Activate step
+	activateResult := r.activateStep(step, stepID, defaultStepLibSource, buildRunResults, isStepLibOfflineMode)
+	if activateResult.Err != nil {
+		return newActivateAndRunStepResult(activateResult.Step, activateResult.StepInfoPtr, models.StepRunStatusCodePreparationFailed, 1, activateResult.Err, true, map[string]string{}, nil)
 	}
 
-	ciModeEnv := os.Getenv(configs.CIModeEnvKey)
+	stepInfoPtr := activateResult.StepInfoPtr
+	mergedStep := activateResult.Step
+	stepIDData := activateResult.StepIDData
+	stepDir := activateResult.StepDir
 
-	if ciModeEnv == "true" {
-		return true, nil
-	}
+	//
+	// Run step
+	logStepStarted(stepInfoPtr, mergedStep, stepIDx, stepExecutionID, stepStartTime)
 
-	for _, env := range inventoryEnvironments {
-		key, value, err := env.GetKeyValuePair()
+	// Evaluate run conditions
+	if mergedStep.RunIf != nil && *mergedStep.RunIf != "" {
+		buildFailedEnvs := bitrise.BuildFailedEnvs(buildRunResults.IsBuildFailed())
+		runIfEnvs := append(environments, buildFailedEnvs...)
+		runIfEnvList, err := envman.ConvertToEnvsJSONModel(runIfEnvs, true, false, &envmanEnv.DefaultEnvironmentSource{})
```

Comment:
Previously the env vars were read from the step's Input EnvStore file:

```
envList, err := tools.EnvmanReadEnvList(configs.InputEnvstorePath)
```

Because of this, we had to initialize the Step Input EnvStore file and add the current build env list to the file.

With this change, we can do the Step Input EnvStore file init after evaluating the Step's run conditions (RunIf and IsAlwaysRun) and also all the step env var preparation work could be merged and moved to the new `prepareEnvsForStepRun` function.